aten/src/ATen/native/cuda/TensorShape.cu

```diff
@@ -268,6 +272,277 @@ std::pair<at::Tensor, std::vector<int64_t*>> pack_vecs(
+  auto num_tensors = tensors.size();
+  const auto device = tensors[0].device();
+  int64_t leading_dim = 1;
+  auto first_tensor_sizes = tensors[0].sizes();
+  if (dim > 0) {
+    leading_dim = c10::multiply_integers(first_tensor_sizes.slice(0, dim));
+  }
+  std::vector<int64_t> pad_tensor_chunk_sizes;
+  std::vector<int64_t> num_blocks_per_tensor_chunk;
+  std::vector<int64_t> start_block_idx_per_tensor_chunk{0};
+  std::vector<int64_t> actual_tensor_sizes;
+  std::vector<int64_t> tensor_idx_to_start_tensor_bytes{0};
+  std::vector<int64_t> srcs;
+  pad_tensor_chunk_sizes.reserve(num_tensors);
+  num_blocks_per_tensor_chunk.reserve(num_tensors);
+  start_block_idx_per_tensor_chunk.reserve(num_tensors + 1);
+  actual_tensor_sizes.reserve(num_tensors);
+  tensor_idx_to_start_tensor_bytes.reserve(num_tensors + 1);
+  srcs.reserve(num_tensors);
+  //  block_idx_to_tensor_idx cannot be reserved since the number of blocks is
+  //  data dependent
+  std::vector<int64_t> block_idx_to_tensor_idx;
+  int64_t chunk_size = 0;
+  for (const auto i : c10::irange(num_tensors)) {
+    at::Tensor tensor = tensors[i];
+    srcs.push_back(reinterpret_cast<int64_t>(tensor.data_ptr()));
+    auto sizes = tensor.sizes();
+    const int64_t size_along_dim = sizes[dim];
+    int64_t trailing_numel = 1;
+    if (sizes.size() > (uint64_t)dim + 1) {
+      trailing_numel =
+          c10::multiply_integers(sizes.slice(dim + 1, sizes.size() - dim - 1));
+    }
+    const int64_t pad_size_along_dim =
+        detail::div_up(size_along_dim, num_chunks) * num_chunks;
+    const int64_t pad_tensor_chunk_size =
+        pad_size_along_dim * trailing_numel * dst_elem_size / num_chunks;
+    pad_tensor_chunk_sizes.push_back(pad_tensor_chunk_size);
+    chunk_size += pad_tensor_chunk_size;
+    // Number of blocks required to process this tensor chunk.
+    const int64_t num_blocks =
+        detail::div_up(pad_tensor_chunk_size, detail::BYTES_PER_BLOCK);
+    num_blocks_per_tensor_chunk.push_back(num_blocks);
+    start_block_idx_per_tensor_chunk.push_back(
+        start_block_idx_per_tensor_chunk.back() + num_blocks);
+    block_idx_to_tensor_idx.insert(
+        block_idx_to_tensor_idx.end(), num_blocks, i);
+    tensor_idx_to_start_tensor_bytes.push_back(
+        tensor_idx_to_start_tensor_bytes.back() + pad_tensor_chunk_size);
+    actual_tensor_sizes.push_back(
+        size_along_dim * trailing_numel * src_elem_size);
+  }
+  const int64_t num_blocks_per_chunk = start_block_idx_per_tensor_chunk.back();
+  const int64_t slice_size = num_chunks * chunk_size;
+  auto packed = detail::pack_vecs(
+      {&srcs,
+       &block_idx_to_tensor_idx,
+       &tensor_idx_to_start_tensor_bytes,
+       &start_block_idx_per_tensor_chunk,
+       &actual_tensor_sizes,
+       &pad_tensor_chunk_sizes,
+       &num_blocks_per_tensor_chunk},
+      device);
+  return std::make_tuple(
+      chunk_size, leading_dim, num_blocks_per_chunk, slice_size, packed.second);
```

Comment:
You need to keep the tensor owning the buffer that's holding these vecs alive until you launch the kernel.

Reply:
(I'm not familiar with how the HIP caching allocator's behavior differs from CUDA. But this is my guess of what went wrong)

Reply:
HIP caching allocator behavior is identical to the CUDA one.  So this would be a bug for CUDA too but it hasn't been observed yet.

Reply:
@jeffdaily we see it too in CUDA: https://github.com/pytorch/pytorch/issues/122026